### PR TITLE
[ASTPrinter] Print expressions

### PR DESCRIPTION
**Pull request by louisdh**
Add new `-print-ast-decl` frontend option for only printing declarations,
to match existing behavior.
Some tests want to print the AST, but don't care about expressions.

The existing `-print-ast` option now prints function bodies and expressions.
Not all expressions are printed yet, but most common ones are.

### CODE REVIEW

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [inline lib/AST/ASTPrinter.cpp:4617]: `do while` has been renamed to `repeat while` for a few years now... 🙂

File: test/Sema/struct_equatable_hashable_access.swift
Comment (louisdh) [inline test/Sema/struct_equatable_hashable_access.swift:1]: It might be useful to add tests for the synthesized _bodies_ for `Equatable`/`Hashable`/etc. But this existing test just wants to check the access level. Using the new `-print-ast-decl` makes this the least disruptive.

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4126]: Note you can also iterate it as `for (auto arg : *argList)`. You could even write this as:

```cxx
llvm::interleave(*argList, [&](auto arg) { printArgument(arg); },
                 [&]() { Printer << ", "; });
```

to automatically handle the intermixing of the commas. You could also use `llvm::interleave` in a couple of the below methods, though I don't feel too strongly about it.

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4180]: I *believe* it's guaranteed for a `DictionaryExpr` to have `TupleExpr` elements, each with two elements, so I think you should be able to `cast<TupleExpr>` here, and access the key and value tuple elements directly with `->getElement` later. Really we ought to assert this in `DictionaryExpr`'s constructor (or better yet move it onto a better element representation).

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp]: That makes the code a lot cleaner, thanks!

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp]: Updated to use the simpler iteration. Not using `interleave` for now, since existing code in this file doesn't either. I can update it to use `interleave` if people feel strongly about that, of course.

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:1330]: Does `Options.PrintExprs` need to be checked here vs letting `visit` handle it?

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:1302]: Why remove this? Do we still print the `?` for e.g:

```
func foo(_ x: Int?) {
  switch x {
  case let x?:
    break
  case nil:
    break
  }
}
```

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4367]: I believe just `visit(expr->getSubExpr())` would be sufficient here, could you add a test to make sure a method call on a protocol-typed value gets printed correctly?

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4469]: I think we'll want to print the `as? Ty` here

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4405]: I think we'll want to print the `as! Ty` here

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4253]: I wonder if we could stamp out the same implementation for every `ImplicitConversionExpr` (as I don't believe any of them would require printing logic)? ExprNodes.def would probably have to be updated to use a new macro kind for such expressions, but then we could have e.g:

```cxx
#define IMPLICIT_CONVERSION_EXPR(Id, Parent) \
void PrintAST::visit##Id##Expr(Id *expr) {   \
  visit(expr->getSubExpr());                 \
}
#include "swift/AST/ExprNodes.def"
```

Doesn't have to be done in this PR though.

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4458]: Should we print the generic arguments here?

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [inline lib/AST/ASTPrinter.cpp:4535]: I think we'll still want to print the `return` keyword even if there's no result, e.g:

```swift
func foo() {
  if .random() { return }
}
```

**Review by hamishknight — COMMENTED** [review]
Great to see this being added! I have some more comments, but otherwise this LGTM

File: include/swift/AST/PrintOptions.h
Comment (louisdh) [reply include/swift/AST/PrintOptions.h]: Yes, good idea, fixed!

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp]: Fixed and added test.

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp]: Good catch, fixed and added test.

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp]: Looks like we do need `visit(expr->getExistentialValue())`, but not `visit(expr->getValueProvidingExpr())`.
Added test to verify.

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp:4254]: I wonder if that's worth it for those one-liners. In any case, I don't want to attempt that in this PR.

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp:4467]: I couldn't find an example where this expression is created. 

Generally, I'd like to merge this even if some expressions aren't printed 100% correct yet (there certainly are more parts missing).

Comment (hamishknight) [reply lib/AST/ASTPrinter.cpp:4467]: It's created by the parser for e.g `Array<Int>`, before the types have been resolved. It gets rewritten by the type-checker though, so I don't believe the current frontend printing option will come across it. But if we ever wanted to print the AST prior to type-checking, I think we should print the generic arguments. But yeah, that doesn't need to hold up this PR.

File: test/expr/print/protocol.swift
Comment (hamishknight) [inline test/expr/print/protocol.swift:23]: I don't think we'll want to print access modifiers here

File: test/expr/print/switch.swift
Comment (hamishknight) [inline test/expr/print/switch.swift:37]: Looks like the `case` keyword was dropped here

**Review by hamishknight — APPROVED** [review]
Couple of extra comments, but I'm happy for them to be done in a follow-up

File: lib/AST/ASTPrinter.cpp
Comment (louisdh) [reply lib/AST/ASTPrinter.cpp]: `visitCaseStmt` calls this, which doesn't check for `Options.PrintExprs`, so I thought it was safest to conditionalize it here.

With that said, I removed the condition and all tests pass, so maybe it's safe.

File: test/expr/print/protocol.swift
Comment (louisdh) [reply test/expr/print/protocol.swift:23]: It's really only used for debugging right now. If this was to be presented to a user, we'd want to filter this so it becomes:
```
let conditional = a as? Archivable
```
(removing `@_hasInitialValue private` and the explicit type annotation, if possible)

File: lib/AST/ASTPrinter.cpp
Comment (hamishknight) [reply lib/AST/ASTPrinter.cpp]: Isn't the `visit(expr)` call guaranteed to check it?

File: test/expr/print/protocol.swift
Comment (hamishknight) [reply test/expr/print/protocol.swift:23]: Right yeah, my main concern was just that e.g:

```
func foo() {
  private let x = 0
}
```

isn't valid code

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by louisdh** [comment]
@swift-ci please test

**Comment by louisdh** [comment]
@swift-ci please test

**Pull request merged**
